src/include/api.h

```diff
@@ -253,6 +253,15 @@
     if (F_ISSET(cur, WT_CURSTD_CACHED))                                                    \
     WT_ERR(__wt_cursor_cached(cur))
 
+#define API_RETRYABLE(s) do {
+
+#define API_RETRYABLE_END(s, ret)                                                                \
+    if ((ret) != WT_ROLLBACK || F_ISSET((s)->txn, WT_TXN_RUNNING) || (s)->api_call_counter != 1) \
```

Comment:
Should we worry about a situation where we keep hitting `WT_ROLLBACK`? I.e., should there be a max number of retries after which we give up and return `WT_ROLLBACK` to the application.

It seems awkward. Then the application would probably retry and we wouldn't make progress and applications would still need the retry logic that this PR is supposed to change.

So this is really a question about whether there is enough of a chance to hit an infinite `WT_ROLLBACK` scenario that we should be worried about it.

Reply:
I don't know the answer to your final question.  I'll note that the same issue is in play for the existing retry loop around the "auto" transactions that happen for write operations.

Reply:
SGTM. If the transactions around writes seem more likely to generate rollbacks. And if we're not getting in trouble there, I'm comfortable assuming it won't be a problem for reads.